data/src/main/java/com/microsoft/azure/kusto/data/BaseClient.java

```diff
@@ -0,0 +1,185 @@
+package com.microsoft.azure.kusto.data;
+
+import com.azure.core.http.*;
+import com.azure.core.util.Context;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.microsoft.azure.kusto.data.exceptions.*;
+import com.microsoft.azure.kusto.data.http.HttpRequestBuilder;
+import com.microsoft.azure.kusto.data.http.HttpStatus;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.invoke.MethodHandles;
+import java.util.Objects;
+import java.util.Optional;
+
+public abstract class BaseClient implements Client, StreamingClient {
+
+    private static final int MAX_REDIRECT_COUNT = 1;
+
+    // Make logger available to implementations
+    protected static final Logger LOGGER = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final HttpClient httpClient;
+
+    public BaseClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    protected String post(HttpRequest request) throws DataServiceException {
+        // Execute and get the response
+        try (HttpResponse response = httpClient.sendSync(request, Context.NONE)) {
+            return processResponseBody(response);
+        }
+    }
+
+    protected Mono<String> postAsync(HttpRequest request) {
+        // Execute and get the response
+        return httpClient.send(request)
+                .onErrorContinue((err, src) -> LOGGER.error("Error coming from src {}", src, err))
```

Comment:
so this is not what you need here . You probably wanted to use doOnError(). and you only need to do it once as its going to propagate all they way

Reply:
For some reason reactor is freezing, so I put in onErrorContinue's in the hopes that an error would propagate and I would see what the issue is. Commented the async implementations out for now because they are not working yet.